packages/nimble/inst/CppCode/nimOptim.cpp

```diff
@@ -0,0 +1,51 @@
+#include <R_ext/Applic.h>
+#include <algorithm>
+#include <cmath>
+#include <limits>
+#include <nimble/nimOptim.h>
+
+// This wraps a NimObjectiveFn as an R `optimfn` type
+//   typedef double optimfn(int, double *, void *);
+// defined in
+//   https://svn.r-project.org/R/trunk/src/include/R_ext/Applic.h
+static double optimfn_wrapper(int n, double* par, void* fn) {
+    NimArr<1, double> nim_par;
+    nim_par.setSize(n);
+    std::copy(par, par + n, nim_par.getPtr());
+    return ((NimObjectiveFn*)fn)(nim_par);
+}
+
+// This attempts to match the behavior of optim() defined in the documentation
+//   https://stat.ethz.ch/R-manual/R-devel/library/stats/html/optim.html
+// and in the reference implementation
+//   https://svn.r-project.org/R/trunk/src/library/stats/R/optim.R
+//   https://svn.r-project.org/R/trunk/src/library/stats/src/optim.c
+//   https://svn.r-project.org/R/trunk/src/include/R_ext/Applic.h
+nimSmartPtr<OptimResultNimbleList> nimOptim(NimArr<1, double>& par,
```

Comment:
We could probably be a bit more general if we take the parameter vector as a NimArrBase<double>, meaning we'll access high-dimensional objects in column-major order as vectors.  One could argue it is preferable to always require it to be 1-dimensional for clarity.  Not very important - just mentioning it.